### PR TITLE
chore: Use GitHub Actions to build Docker Images to allow publishing Windows Images

### PR DESCRIPTION
**Pull request by lippertmarkus**
As mentioned in #2747 Docker Hub Automated Build Hooks don't allow to build Windows images. As v2.9.0 comes with Windows support for argoexec and as you migrated CI to GitHub Actions also, it would make sense to use GitHub Actions also for building the Docker images.

I added a workflow file which builds and pushes the images for Linux (workflow-controller, argoexec, argocli) and Windows (argoexec only) whenever a new tag `v*` is pushed. I used the existing Docker Hub Hooks (https://github.com/argoproj/argo/blob/master/hooks/build) as a base for the scripts. The jobs building and pushing the images are appending `-linux` or `-windows` to the tag, a third job is creating a Manifest and pushes it to allow for multiarch. Example:

1. Tag `v2.9.0` is pushed
1. Job `build-linux` is building and pushing `argoproj/workflow-controller:v2.9.0-linux`, `argoproj/argoexec:v2.9.0-linux` `argoproj/argocli:v2.9.0-linux`
1. Job `build-windows` is building and pushing `argoproj/argoexec:v2.9.0-windows`
1. Job `push-images` is creating and pushing the following image manifest:
   - `argoproj/workflow-controller:v2.9.0` referencing `argoproj/workflow-controller:v2.9.0-linux`
   - `argoproj/argoexec:v2.9.0` referencing `argoproj/argoexec:v2.9.0-linux` and `argoproj/argoexec:v2.9.0-windows` (multiarch)
   - `argoproj/argocli:v2.9.0` referencing `argoproj/argocli:v2.9.0-linux`

You can find an example run for a tag `vtest17` here: https://github.com/lippertmarkus/argo/runs/804344322
Resulting image, e.g. for argoexec: https://hub.docker.com/repository/docker/lippertmarkus/argoexec/tags

You only need to add the following two secrets like described [in the GitHub Actions Help](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#creating-encrypted-secrets-for-a-repository) before merging:
- `DOCKERIO_USERNAME` = `argoproj`
- `DOCKERIO_PASSWORD` = `<a-docker-hub-pat-token>`

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or **(c) this is a chore**.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged. 
  - not needed
* [x] My builds are green. Try syncing with master if they are not.

### CODE REVIEW

**Comment by alexec** [comment]
Let me get this right, would this save us building this locally? Just tag and push tag?

**Comment by lippertmarkus** [comment]
Yes, correct.

**Comment by lippertmarkus** [comment]
What do you think? Would also help to make the builds more reproducable.

**Comment by alexec** [comment]
We still need to build binaries. Any thoughts on how that might be possible? Maybe another PR?

**Comment by lippertmarkus** [comment]
This would be possible with the following steps:
- Automatically create a release: https://github.com/marketplace/actions/create-a-release
- Upload assets to the newly created release: https://github.com/marketplace/actions/upload-a-release-asset (binaries can be build with another job)

Afterwards you could add your release description manually or you also automate that.

Could look into that in another PR, but this PR would already help a lot with getting "official" Windows Images for argoexec :)

**Comment by lippertmarkus** [comment]
any updates on this PR?

File: .github/workflows/docker-image.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
+name: Docker Image CI
+
+on:
+  push:
+    tags:
+      - v*
```
Comment (alexec) [inline .github/workflows/docker-image.yml:6]: should also include `master`

Comment (lippertmarkus) [reply .github/workflows/docker-image.yml:6]: Ok, to be clear: Every commit on `master` should push a new image with `latest` tag?

**Comment by lippertmarkus** [comment]
For pushing new images we would always need to store credentials. Those are stored securely by GitHub (symmetric encryption, see [here](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#about-encrypted-secrets)). This is done this way by a lot of other projects on GitHub so there shouldn't be a problem.

And you don't use your normal Docker Hub password, you would instead create a new [Access Token on Docker Hub](https://docs.docker.com/docker-hub/access-tokens/) and use this for authentication. This token can be revoked at anytime if needed.

File: .github/workflows/docker-image.yml (hunk 1 above)
Comment (alexec) [reply .github/workflows/docker-image.yml:6]: correct

**Comment by alexec** [comment]
I have now set-up the secrets for this.

**Comment by lippertmarkus** [comment]
Thanks! I added the master branch and tested again. You can view an example run [here](https://github.com/lippertmarkus/argo/actions/runs/162922625) and resulting `latest` docker images e.g. for argoexec [here](https://hub.docker.com/repository/docker/lippertmarkus/argoexec/tags?page=1&ordering=last_updated).

**Comment by alexec** [comment]
Lets see if this works.

**Comment by alexec** [comment]
https://github.com/argoproj/argo/actions/runs/163542666

But where are the images?

**Comment by lippertmarkus** [comment]
hmm, push seems to be successful. What did you specify as Docker username?

**Comment by alexec** [comment]
my username: `alexcollinsintuit`

**Comment by alexec** [comment]
ah

https://hub.docker.com/repository/registry-1.docker.io/alexcollinsintuit/argocli/tags?page=1

**Comment by alexec** [comment]
can you try them out to see if they work?

**Comment by lippertmarkus** [comment]
I tested the images, they work fine. We need to use a different repository than the Docker Hub username you log-in with. I created a new PR: #3442.

**Pull request merged**